### ISSUE TITLE
MQTT Subscribe succeeds, Publish Fails

### ISSUE THREAD

**Issue by jeffreyb-arcules** [meetecho/janus-gateway#1535]
I built janus 0.6.2 from scratch. I am running a local mosquitto broker without authentication or ssl.

Janus can connect to the broker, subscribe to topics, and receive messages, but, when it tries to publish its response it fails.  The error code is -15, which correspons to invalid MMQTT options. It feels like something in the configuration file might be missing... 

Here is the log, thanks in advance!

[Wed Mar  6 15:46:41 2019] MQTT client has been successfully connected to the broker
[Wed Mar  6 15:46:41 2019] MQTT client has been successfully subscribed to MQTT topic: to-janus
[Wed Mar  6 15:49:15 2019] Receiving Janus API message over MQTT: {"janus":"create","transaction":"5c95d502-483f-4f9b-bc32-4ecbc31e4174"}
[Wed Mar  6 15:49:15 2019] Got a Janus API request from janus.transport.mqtt (0x5603d8ba2ee0)
[Wed Mar  6 15:49:15 2019] Creating new session: 2057862836598528; 0x7f1048001320
[Wed Mar  6 15:49:15 2019] Sending Janus API response to janus.transport.mqtt (0x5603d8ba2ee0)
[Wed Mar  6 15:49:15 2019] Sending Janus API message via MQTT: {
   "janus": "success",
   "transaction": "5c95d502-483f-4f9b-bc32-4ecbc31e4174",
   "data": {
      "id": 2057862836598528
   }
}
[Wed Mar  6 15:49:15 2019] [ERR] [transports/janus_mqtt.c:janus_mqtt_send_message:488] Can't publish to MQTT topic: from-janus, return code: -15

**Comment by jeffreyb-arcules** [comment]
here is the mqtt config:
$ cat etc/janus/janus.transport.mqtt.jcfg
# Configuration of the MQTT additional transport for the Janus API.
general: {
	enabled = true						# Whether the support must be enabled
	json = "indented"						# Whether the JSON messages should be indented (default),
										# plain (no indentation) or compact (no indentation and no spaces)

	url = "tcp://localhost:1883"		# The connection URL of the MQTT broker: if you want
										# to use SSL, make sure you type ssl:// instead of tcp://,
										# and that you configure the SSL settings below
	client_id = "janus-edge"				# Client identifier
	#username = "tu1"					# Username to use to authenticate, if needed
	#password = "password"					# Password to use to authenticate, if needed
	keep_alive_interval = 20			# Keep connection for N seconds
	cleansession = 0					# Clean session flag
	#disconnect_timeout = 100			# Milliseconds to wait before destroying client
	subscribe_topic = "to-janus"		# Topic for incoming messages
	subscribe_qos = 1					# QoS for incoming messages
	publish_topic = "from-janus"		# Topic for outgoing messages
	publish_qos = 1					# QoS for outgoing messages

	ssl_enabled = false					# Whether ssl support must be enabled
	verify_peer = false					# Whether peer verification must be enabled

	# Certificates to use when SSL support is enabled, if needed
	#cacertfile = /path/to/cacert.pem
	#certfile = /path/to/cert.pem
	#keyfile = /path/to/key.pem

}

admin: {
	admin_enabled = false				# Whether the support must be enabl
	subscribe_topic = "to-janus-admin"	# Topic for incoming admin messages
	#subscribe_qos = 1					# QoS for incoming admin messages
	publish_topic = "from-janus-admin"	# Topic for outgoing admin messages
	#publish_qos = 1					# QoS for outgoing admin messages
}

**Comment by lminiero** [comment]
Sorry but I'm not familiar with MQTT, and I have no idea of what -15 as an error means. You may want to investigate that and let us know, so that we can check if it's a bug in our code or something you're doing wrong.

**Comment by jeffreyb-arcules** [comment]
Thanks for your response, I did look up error 15 and pasted in the original post. Perhaps it was too wordy:

> The error code is -15, which correspons to invalid MMQTT options. It feels like something in the configuration file might be missing...

**Comment by lminiero** [comment]
I didn't change anything in the plugin code, and I tried using the same test client I used successfully months ago, but I'm getting the same error too now:

```
[ERR] [transports/janus_mqtt.c:janus_mqtt_send_message:488] Can't publish to MQTT topic: from-janus, return code: -15
```

Unfortunately I'm not that familiar with MQTT, so no idea if anything changed in Mosquitto to cause this issue that never happened before. Pinging @manifest and @oej as they may know more.

**Comment by lminiero** [comment]
The MQTT event handler is affected too:

```
[WARN] FAILURE: MQTT EVH message propably not sent to topic /janus/events/transports on tcp://localhost:1883. Result -15
[WARN] Can't publish to MQTT topic: /janus/events, return code: -15
```

**Comment by lminiero** [comment]
Looks like we might have to explicitly specify the version field in the `MQTTAsync_responseOptions`, as this is what seems to trigger a `MQTTASYNC_BAD_MQTT_OPTION` error when sending a message with paho.mqtt.c:

		if (m->c->MQTTVersion >= MQTTVERSION_5)
		{
			if (response->struct_version == 0 || response->onFailure || response->onSuccess)
				rc = MQTTASYNC_BAD_MQTT_OPTION;
		}
		else if (m->c->MQTTVersion < MQTTVERSION_5)
		{
			if (response->struct_version >= 1 && (response->onFailure5 || response->onSuccess5))
				rc = MQTTASYNC_BAD_MQTT_OPTION;
		}

My guess is that, since we're not setting it, we might be sending 0, so 1 might fix it. Anyway, I don't know if that will be enough, and if we'll have to fix something else in the code too.

**Comment by lminiero** [comment]
Yep, that was the issue, but for a different reason: we were not setting the `MQTTAsync_responseOptions` instance to 0 after declaring it, and so struct version was very likely different than 0 because we were not setting it. Fixing as I type.
